datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java

```diff
@@ -1349,10 +1349,18 @@ private void handleLeaderDoAssignment(boolean cleanUpOrphanNodes) {
         failedStreams = unclaimedTokens.keySet();
         Set<String> hosts = unclaimedTokens.values().stream().flatMap(List::stream).
```

Comment:
out-of-curiousity: Is there a reason why entry into this block is predicated on `getNumUnclaimedTokensForDatastreams` (time bounded) but requires `getUnclaimedAssignmentTokensForDatastreams` (unbounded)? These two requests could have different results. Any non-zero list is an equivalent test and an error could be interpreted as "success not confirmed". Since there is no recovery here, it seems or important to make a best-effort attempt to verify no unclaimed tokens are present and honor the time constraint.

I realize this was added in #922  and not here, but this question wasn't addressed in any of comments there.

Reply:
Can you explain how `getNumUnclaimedTokensForDatastreams` is time bounded? What do you mean by time bounded here?

The reason I have `getNumUnclaimedTokensForDatastreams` in the if statement condition is because it is faster. It doesn't involve reading the zookeeper data and deserializing it. I understand that the results of `getNumUnclaimedTokensForDatastreams` and `getUnclaimedAssignmentTokensForDatastreams` can be inconsistent in the edge cases. By the time the leader calls `getUnclaimedAssignmentTokensForDatastreams` their number may be lower than what was earlier reported. It can even be zero. This, however, won't affect the outcome.

Reply:
> Can you explain how `getNumUnclaimedTokensForDatastreams` is time bounded? What do you mean by time bounded here?

I was referring to the `PollUtils` wrapper and timeout. This wrapped request will be interrupted if incomplete while the result from `getUnclaimedAssignmentTokensForDatastreams` will not.
```java
if (_config.getEnableAssignmentTokens() &&
        !PollUtils.poll(() -> _adapter.getNumUnclaimedTokensForDatastreams(stoppingDatastreamGroups) == 0,
        STOP_PROPAGATION_RETRY_MS, _config.getStopPropagationTimeout())) {
    Map<String, List<AssignmentToken>> unclaimedTokens =
            _adapter.getUnclaimedAssignmentTokensForDatastreams(stoppingDatastreamGroups);
````

----

I agree that the `..Num..` is faster and do not believe there is anything incorrect with either method; but I am not sure the faster option is valuable though. For the happy path, no unassigned tokens, the request will be fast, because `zkclient.getChildren()` will be be empty. In the failure case, identifying the failures, unclaimed tokens, is required. If queried with `...Num...`, a second query is necessary to retrieve the unassigned tokens. However, the second query may be have a different result from the first. It may be better to query once and capture state clearly if we plan on debugging unclaimed token issues. 

To reclaim some performance, the deserialization can be deferred and return a `List<String>`. The error handling block or `revokeUnclaimedAssignmentTokens` can convert to `AssignmentToken`s. (I don't have a preference for deferring deserialization.)

Reply:
@ehoner `getUnclaimedAssignmentTokens` might return fewer tokens than `..Num..`, because the invocation order is `..Num..` followed by `getUnclaimed...`. So the fact that these are two separate methods won't change the overall outcome. At the end of the day, the remaining tokens (whatever is returned by the last call) will get processed.

I would want to avoid exposing List<String> and leaking serialized tokens outside of the `ZkAdapter`. Classes that use ZkAdapter shouldn't worry about ser/de aspect of ZK storage.

Reply:
Discussed offline. 
tldr; The issue of detecting different results between the two queries isn't of concern.

Differing results between the first request to `..Num..` and token retrieval does not produce any material behavior changes. If unclaimed token(s) were detected by the first query, but claimed before the second, the only actions are logging and `revokeUnclaimedTokens`. The log statement will show an empty set and `revoke..` will be a noop. More importantly, though no person will be signaled to investigate and no "corrective" actions would be required.